torch/distributed/fsdp/_utils.py

```diff
@@ -1,47 +1,69 @@
 
 
+# Save a global mapping from module to its input tensor dtype to be populated
+# during the forward pre-hook and consumed in the forward post-hook when
+# overriding a module's mixed precision
+# NOTE: We currently take the last input tensor's dtype in the case of multiple
+# floating-point input tensors, which may be incorrect. However, since there is
+# not a 1:1 correspondence between input and output tensors, we must use *some*
+# heuristic like this to predict the desired output dtype.
+_MODULE_TO_INP_DTYPE: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+
+
 def _override_module_mixed_precision(
     root: torch.nn.Module,
-    module_cls_to_override: Type[torch.nn.Module],
+    module_classes_to_override: Iterable[Type[nn.Module]],
     wrap_override_dict: Dict[str, Any] = {"mixed_precision": None},  # noqa: B006
-):
+) -> Set[Type[nn.Module]]:
+    module_classes_to_override = tuple(set(module_classes_to_override))
+    # Return a set of the actually overridden module classes
+    overridden_module_classes: Set[Type[nn.Module]] = set()
     for mod in root.modules():
-        if isinstance(mod, module_cls_to_override):
+        if isinstance(mod, module_classes_to_override):
+            overridden_module_classes.add(type(mod))
             mod._wrap_overrides = wrap_override_dict  # type: ignore[assignment]
             # TODO: We need to run this mixed precision ignored module in fp32,
             # but ensure subsequent modules, that may possibly be running with
             # mixed precision, still receive the appropriate precision inputs
             # without user having to adjust mixed precision config too much.
             # As a result, we attach pre and post forward hooks to up / down
             # cast. We should revisit this design.
-            old_dtype = None
 
-            def cast_fn(dtype, x: torch.Tensor) -> torch.Tensor:
+            def cast_fn(
+                dtype: torch.dtype, module: nn.Module, x: torch.Tensor
+            ) -> torch.Tensor:
                 if not torch.is_floating_point(x) or x.dtype == dtype:
                     return x
-                nonlocal old_dtype
-                old_dtype = x.dtype
+                _MODULE_TO_INP_DTYPE[module] = x.dtype
                 return x.to(dtype)
 
             def forward_pre_hook(module, args):
-                return _apply_to_tensors(partial(cast_fn, torch.float32), args)
+                return _apply_to_tensors(partial(cast_fn, torch.float32, module), args)
 
             def forward_post_hook(module, args, output):
-                nonlocal old_dtype
-                if old_dtype is not None:
-                    return _apply_to_tensors(partial(cast_fn, old_dtype), output)
+                # NOTE: If the forward did not have any floating-point tensors,
+                # then the dtype will not be set for this module, and we do not
+                # upcast the dtype.
+                if module in _MODULE_TO_INP_DTYPE:
```

Comment:
so _MODULE_TO_INP_DTYPE generalizes old_dtype to be on a per-module basis?

Reply:
Yep!